### PR TITLE
cgit 1.1 (new formula)

### PR DESCRIPTION
**Pull request by FranklinYu**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

See also [previous discussion](https://discourse.brew.sh/t/new-formula-cgit/2504). Concerns:

1.  The installer writes something in `CGIT_DATA_PATH`. When upgrading, should it simply overwrite, or do something similar to `etc.install`?
2.  Shall I note this in caveats section?
3.  The test passes and returns 0, but this is not how CGI works. Current test prints “404 Not found”. Shall I add all the CGI variables to get a “200 Ok”?

### CODE REVIEW

**Comment by franklinyu** [comment]
What does the error means?

> * Formulae should not have [:recommended] dependencies.
> * Formulae should not have [:optional] dependencies.

Are we deprecating optional and recommended dependencies?

**Comment by ilovezfs** [comment]
@fran6co options aren't allowed on new formulae.

**Comment by fran6co** [comment]
@ilovezfs I think you got the wrong handle

**Comment by ilovezfs** [comment]
@fran6co yes sorry 😊

**Comment by ilovezfs** [comment]
@franklinyu options aren't allowed on new formulae.

**Comment by franklinyu** [comment]
@ilovezfs Ok. Now that maintainer has full control on what's included, is there any guideline about what to include?

**Comment by ilovezfs** [comment]
@franklinyu yes, we want to strive for one-size-fits-all configuration, so whatever makes the formula the most useful, erring on the side of upstream defaults and fewer dependencies if it's a close call. In practice that usually means requiring what you might previously have made `=> :recommended` and just dropping what you might previously have made `=> :optional`. Sometimes it means requiring what you would have made `=> :optional` if you think it will make the formula significantly more useful without being overly burdensome.

File: Formula/cgit.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class Cgit < Formula
+  desc "Hyperfast web frontend for Git repositories written in C"
+  homepage "https://git.zx2c4.com/cgit/"
+  url "https://git.zx2c4.com/cgit/snapshot/cgit-1.1.tar.xz"
+  sha256 "0889af29be15fc981481caa09579f982b9740fe9fd2860ab87dff286f4635890"
+  depends_on "openssl@1.1"
```
Comment (ilovezfs) [inline Formula/cgit.rb]: please use 1.0 if it works

**Comment by franklinyu** [comment]
Am I blocking anything? Or anything I can help?

File: Formula/cgit.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class Cgit < Formula
+  desc "Hyperfast web frontend for Git repositories written in C"
+  homepage "https://git.zx2c4.com/cgit/"
+  url "https://git.zx2c4.com/cgit/snapshot/cgit-1.1.tar.xz"
+  sha256 "0889af29be15fc981481caa09579f982b9740fe9fd2860ab87dff286f4635890"
+  depends_on "gettext"
```
Comment (DomT4) [inline Formula/cgit.rb]: Needs an empty line between `sha256` and the first `depends_on`.

File: Formula/cgit.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  depends_on "openssl"
+
+  def install
+    system "make", "get-git"
+    system "make",
```
Comment (DomT4) [inline Formula/cgit.rb]: `system "make", "prefix=#{prefix}",` and then everything aligned with the `prefix` part is usually the style used.

File: Formula/cgit.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      "prefix=#{prefix}",
+      "CGIT_SCRIPT_PATH=#{pkgshare}",
+      "CGIT_DATA_PATH=#{var}/www/htdocs/cgit",
+      "CGIT_CONFIG=#{etc}/cgitrc",
+      "CACHE_ROOT=#{var}/cache/cgit",
+      "install"
+  end
+
+  test do
+    system opt_pkgshare/"cgit.cgi"
```
Comment (DomT4) [inline Formula/cgit.rb]: `pkgshare`. Don't need the `opt` bit here.

**Comment by ilovezfs** [comment]
>The installer writes something in CGIT_DATA_PATH. When upgrading, should it simply overwrite, or do something similar to etc.install

How does the installer handle it on its own? I assume `make install` doesn't just overwrite the user data.

>The test passes and returns 0, but this is not how CGI works. Current test prints “404 Not found”. Shall I add all the CGI variables to get a “200 Ok”?

Ideally yes.

**Comment by franklinyu** [comment]
> How does the installer handle it on its own? I assume `make install` doesn't just overwrite the user data.

It simply overwrites. Does it mean that those files are not user data?

File: Formula/cgit.rb (hunk 3 above)
Comment (franklinyu) [reply Formula/cgit.rb]: Those variables do not change when `prefix` changes. I have to manually specify them.

**Comment by ilovezfs** [comment]
It's unlikely that user data would be overwritten by `make install`. What files does it write to CGIT_DATA_PATH?

**Comment by franklinyu** [comment]
* `cgit.css`
* `cgit.png`
* `favicon.ico`
* `robots.txt`

File: Formula/cgit.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+class Cgit < Formula
+  desc "Hyperfast web frontend for Git repositories written in C"
+  homepage "https://git.zx2c4.com/cgit/"
+  url "https://git.zx2c4.com/cgit/snapshot/cgit-1.1.tar.xz"
+  sha256 "0889af29be15fc981481caa09579f982b9740fe9fd2860ab87dff286f4635890"
+  depends_on "gettext"
+  depends_on "openssl"
+
+  def install
+    system "make", "get-git"
```
Comment (ilovezfs) [inline Formula/cgit.rb]: should use a resource block instead of letting build run curl

File: Formula/cgit.rb (hunk 3 above)
Comment (DomT4) [reply Formula/cgit.rb]: I just mean stylistically this is the preferred way of writing this sort of thing _usually_:

```ruby
system "make", "prefix=#{prefix}",
               "CGIT_SCRIPT_PATH=#{pkgshare}"
```
and so on.

File: Formula/cgit.rb (hunk 5 above)
Comment (franklinyu) [reply Formula/cgit.rb]: Shall I hard-code the Git URL, or parse the Makefile for it?

Comment (franklinyu) [reply Formula/cgit.rb]: Wait, the resource is unpacked in another directory. The Makefile assumes that Git is unpacked in `./git/`. How do I symlink it to the resource directory? I don’t see anything like `Resource#path`.

**Comment by franklinyu** [comment]
I will get proper CGI environment later. New to CGI here.

**Comment by franklinyu** [comment]
CGI variables do not seem to be the issue; the page says "error: no Git repositories". Shall I add a file to `$PREFIX/etc/cgitrc` and also create an empty Git repository for the testing?

**Comment by franklinyu** [comment]
Any ideas?

File: Formula/cgit.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  depends_on "gettext"
+  depends_on "openssl"
+
+  resource "git" do
+    url "https://www.kernel.org/pub/software/scm/git/git-2.18.0.tar.gz"
+    sha256 "94faf2c0b02a7920b0b46f4961d8e9cad08e81418614102898a55f980fa3e7e4"
+  end
+
+  def install
+    resource("git").stage(Pathname.new(pwd)/"git")
```
Comment (DomT4) [inline Formula/cgit.rb]: I think `Pathname.new(pwd)` here is just `buildpath`, no?

**Comment by DomT4** [comment]
> Shall I add a file to $PREFIX/etc/cgitrc

Any chance upstream provide an example one of these in the tarball?

> and also create an empty Git repository for the testing?

If it's necessary that's not an issue. See something like `git-cal.rb` if you need an example.

**Comment by franklinyu** [comment]
> > Shall I add a file to `$PREFIX/etc/cgitrc`
>
> Any chance upstream provide an example one of these in the tarball?

There is one at the bottom of <https://git.zx2c4.com/cgit/tree/cgitrc.5.txt>. If there is only one repository, it can be as simple as

```text
repo.url=test
repo.path=/tmp/test-git
repo.desc=the master foo repository
repo.owner=fooman@example.com
```

I’m thinking about how to remove the file after test.

**Comment by franklinyu** [comment]
If the `test` block has several `system` call, some of them failed (let’s say the second among 6 failed), would Homebrew continue to run next `system`, or would it throw and stop the test? For example

```ruby
test do
  system "test-suite-a"
  system "test-suite-b fail"
  system "test-suite-c"
  system "test-suite-d"
end
```

**Comment by franklinyu** [comment]
Is there any suggestion about accessing `etc/"cgitrc"`? I cannot access it during test, due to Homebrew sandbox.

**Comment by claui** [comment]
@franklinyu cgit offers a [temporary override at runtime](https://git.zx2c4.com/cgit/commit/?id=51f6547879e2516b9850b0b3c51df044c16a8a52) via the `CGIT_CONFIG` environment variable, which is probably good enough for testing:

```ruby
test do
  (testpath/"cgitrc").write("root-desc=Brewed with love by Homebrew\n")
  ENV["CGIT_CONFIG"] = testpath/"cgitrc"
  assert_match "with love", shell_output("#{pkgshare}/cgit.cgi")
end
```

**Comment by franklinyu** [comment]
This is ready for review! 😄

**Comment by franklinyu** [comment]
Any news? This is getting stale.

File: Formula/cgit.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  desc "Hyperfast web frontend for Git repositories written in C"
+  homepage "https://git.zx2c4.com/cgit/"
+  url "https://git.zx2c4.com/cgit/snapshot/cgit-1.2.tar.xz"
+  sha256 "d5622ba7c3eb87de402da6fe3a092219546a917ff627966e99c9f18ba32a4832"
+
+  depends_on "gettext"
+  depends_on "openssl"
+
+  resource "git" do
+    url "https://www.kernel.org/pub/software/scm/git/git-2.18.0.tar.gz"
```
Comment (fxcoudert) [inline Formula/cgit.rb]: Can you update to the latest version?
Other than that, it looks OK to merge.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/cgit.rb (hunk 7 above)
Comment (franklinyu) [reply Formula/cgit.rb]: @fxcoudert The CGit requires fixed version of Git to work. See https://git.zx2c4.com/cgit/tree/Makefile?h=v1.2#n17

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @franklinyu for the pull request and for your patience!

**Pull request closed without merging**